src/Service/ProducerEventProcessor.cpp

```diff
@@ -0,0 +1,371 @@
+        {producer_id, mutable_marker.text_key()});
+    CHECK(it != producer_interned_string_id_to_client_string_id_.end());
+    mutable_marker.set_text_key(it->second);
+  }
+
+  ClientCaptureEvent event;
+  *event.mutable_gpu_queue_submission() = std::move(*gpu_queue_submission);
+  capture_event_buffer_->AddEvent(std::move(event));
+}
+
+void ProducerEventProcessorImpl::ProcessFullCallstackSample(
+    FullCallstackSample* full_callstack_sample) {
+  Callstack* callstack = full_callstack_sample->mutable_callstack();
+  auto [callstack_id, assigned] =
+      callstack_cache_.GetOrAssignId({callstack->pcs().begin(), callstack->pcs().end()});
+
+  if (assigned) {
+    ClientCaptureEvent interned_callstack_event;
+    interned_callstack_event.mutable_interned_callstack()->set_key(callstack_id);
+    *interned_callstack_event.mutable_interned_callstack()->mutable_intern() =
+        std::move(*callstack);
+    capture_event_buffer_->AddEvent(std::move(interned_callstack_event));
+  }
+
+  ClientCaptureEvent callstack_sample_event;
+  CallstackSample* callstack_sample = callstack_sample_event.mutable_callstack_sample();
+  callstack_sample->set_pid(full_callstack_sample->pid());
+  callstack_sample->set_tid(full_callstack_sample->tid());
+  callstack_sample->set_timestamp_ns(full_callstack_sample->timestamp_ns());
+  callstack_sample->set_callstack_id(callstack_id);
+  capture_event_buffer_->AddEvent(std::move(callstack_sample_event));
+}
+
+void ProducerEventProcessorImpl::ProcessInternedCallstack(uint64_t producer_id,
+                                                          InternedCallstack* interned_callstack) {
+  CHECK(!producer_interned_callstack_id_to_client_callstack_id_.contains(
+      {producer_id, interned_callstack->key()}));
+
+  std::vector<uint64_t> callstack_data{interned_callstack->intern().pcs().begin(),
+                                       interned_callstack->intern().pcs().end()};
+  auto [interned_callstack_id, assigned] = callstack_cache_.GetOrAssignId(callstack_data);
+
+  producer_interned_callstack_id_to_client_callstack_id_.insert_or_assign(
+      {producer_id, interned_callstack->key()}, interned_callstack_id);
+
+  if (!assigned) {
+    return;
+  }
+
+  // If this is first time we see it -> send it over with client_id
+  interned_callstack->set_key(interned_callstack_id);
+  ClientCaptureEvent event;
+  *event.mutable_interned_callstack() = std::move(*interned_callstack);
+  capture_event_buffer_->AddEvent(std::move(event));
+}
+
+void ProducerEventProcessorImpl::ProcessInternedCallstackSample(uint64_t producer_id,
+                                                                CallstackSample* callstack_sample) {
+  // translate producer id to client id
+  auto it = producer_interned_callstack_id_to_client_callstack_id_.find(
+      {producer_id, callstack_sample->callstack_id()});
+  CHECK(it != producer_interned_callstack_id_to_client_callstack_id_.end());
```

Comment:
Should we crash if something went wrong in the producer (possibly a different process) and didn't send the intern in advance?

Reply:
I do not think we can produce valid capture data in this case.

Reply:
Well, we could just drop the callstack and move on. I know a crash is effectively the only way to see this problem, but as we might depend on external processes it feels wrong to crash.

Reply:
tbh I still would prefer a crash here, since this is something that is caused by programming error even if was in a different process, but if you feel strongly about this, we can log, ignore and hope someone reads the service logs.

Reply:
having thought about this a little bit more, I now think the better way here is to add a error. If this happens capture stops and returns a error message to user. This way we do not throw capture data up to this point away and user has an reason of what exactly went wrong. We might need producer name here so that error message has more context. I will create a bug.

Reply:
added todo:

Reply:
Alright, it seems that we disagree a bit on this, for me even stopping the capture is kind of wrong if one process misbehaves. But it's ok, at least we are tracking it.